### PR TITLE
Add disk always search mode to allow for cluster configuration of Nuster

### PR DESCRIPTION
**Pull request by HugoS99**
This allows for Nuster to always check disk even if the cached file isn't in memory to allow for a configuration where multiple instances of Nuster are running simultaneously with shared cache on disk.

TO-DO
Check if delete code is operational as the file might disappear and still be on memory in a particular node

### CODE REVIEW

File: include/haproxy/global-t.h [hunk 1]
```diff
@@ -194,7 +194,8 @@ struct global {
 			int disk_loader;                 /* the number of files load once */
 			int disk_saver;                  /* the number of entries checked once for persist_async */
 			int clean_temp;                  /* clean temp file or not */
-
+			int disk_always_read;
```
Comment (jiangwenyuan) [inline include/haproxy/global-t.h:197]: It's a little bit confusing, as this one only affects whether or not to check the existence of the cache file on disk if disk-load is done. Do you have other candidates for the var?

File: include/haproxy/global-t.h [hunk 2]
```diff
@@ -194,7 +194,8 @@ struct global {
...
+
```
Comment (jiangwenyuan) [inline include/haproxy/global-t.h:198]: Should delete this line.

File: src/haproxy.c [hunk 3]
```diff
@@ -194,6 +194,7 @@ struct global global = {
 			.disk_loader  = NST_DEFAULT_DISK_LOADER,
 			.disk_saver   = NST_DEFAULT_DISK_SAVER,
 			.clean_temp   = NST_STATUS_OFF,
+			.disk_always_read = NST_STATUS_OFF,
```
Comment (jiangwenyuan) [inline src/haproxy.c:197]: The `=` should be aligned.

File: src/nuster/parser.c [hunk 4]
```diff
@@ -669,6 +669,37 @@ nuster_parse_global_cache(const char *file, int line, char **args) {
             continue;
         }
 
+            if(!strcmp(args[cur_arg], "disk_always_read")) {
```
Comment (jiangwenyuan) [inline src/nuster/parser.c:672]: Please use 4 spaces to indent.

File: src/nuster/parser.c [hunk 5]
```diff
@@ -931,6 +962,36 @@ nuster_parse_global_nosql(const char *file, int line, char **args) {
 
             cur_arg++;
 
+            continue;
+        }
+
+                    if(!strcmp(args[cur_arg], "disk_always_read")) {
```
Comment (jiangwenyuan) [inline src/nuster/parser.c:968]: indent

File: src/nuster/parser.c (hunk 4 above)
Comment (jiangwenyuan) [reply src/nuster/parser.c:672]: And use `-` instenad of `_` for options.

File: src/nuster/store/disk.c [hunk 6]
```diff
@@ -12,7 +12,8 @@
 
 #include <haproxy/tools.h>
 #include <haproxy/global.h>
-
+#include <time.h>
+#include <stdio.h>
```
Comment (jiangwenyuan) [inline src/nuster/store/disk.c:16]: No need

File: src/nuster/store/disk.c [hunk 7]
```diff
@@ -255,7 +256,6 @@ nst_disk_init(nst_disk_t *disk, hpx_ist_t root, nst_shmem_t *shmem, int clean_te
 #ifdef USE_THREAD
 void *nst_disk_load_thread(void *data) {
     nst_core_t  *core = (nst_core_t *)data;
-
```
Comment (jiangwenyuan) [inline src/nuster/store/disk.c:258]: A blank line as separator. Since disk.c is not changed, just reverting should be fine.

**Comment by jiangwenyuan** [comment]
Thanks, added some comments, and also please update the README.md.

File: include/haproxy/global-t.h (hunk 1 above)
Comment (jiangwenyuan) [reply include/haproxy/global-t.h:197]: What do you think of `check_disk_after_load`? a little bit long though

Comment (HugoS99) [reply include/haproxy/global-t.h:197]: How about `alway_check_disk`?

**Comment by HugoS99** [comment]
TO-DO

Still need to update delete purge policy to ensure data is purged

File: include/haproxy/global-t.h (hunk 1 above)
Comment (jiangwenyuan) [reply include/haproxy/global-t.h]: `always` ? Good idea. there's no need to emphasize `after loaded` indeed. Let's go with `always_check_disk`.

File: include/haproxy/global-t.h (hunk 2 above)
Comment (jiangwenyuan) [reply include/haproxy/global-t.h:198]: Line 198 has trailing tabs.

File: src/haproxy.c [hunk 8]
```diff
@@ -185,37 +185,39 @@ struct global global = {
...
+			.clean_temp       = NST_STATUS_OFF,
+			.disk_always_read = NST_STATUS_OFF,
+			.root             = {
 				.ptr  = NULL,
 				.len  = 0,
 			},
 		},
 		.manager = {
-			.status       = NST_STATUS_UNDEFINED,
+			.status           = NST_STATUS_UNDEFINED,
```
Comment (jiangwenyuan) [inline src/haproxy.c:220]: No need for this one, this aligns with purge_method.

File: src/nuster/parser.c [hunk 9]
```diff
@@ -934,6 +965,36 @@ nuster_parse_global_nosql(const char *file, int line, char **args) {
             continue;
         }
 
+            if(!strcmp(args[cur_arg], "disk_always_read")) {
```
Comment (jiangwenyuan) [inline src/nuster/parser.c:968]: This one does not indent well

File: src/nuster/store/disk.c [hunk 10]
```diff
@@ -255,7 +254,7 @@ nst_disk_init(nst_disk_t *disk, hpx_ist_t root, nst_shmem_t *shmem, int clean_te
...
+
```
Comment (jiangwenyuan) [inline src/nuster/store/disk.c:257]: Trailing tab. You can just checkout from master since this file is not changed.

**Comment by jiangwenyuan** [comment]
@HugoS99 Any updates on this?

**Comment by jiangwenyuan** [comment]
Close due to inactivity.

**Comment by jiangwenyuan** [comment]
Added d28c8c8889ca190edbcb494172f1d01f2e1f2366

**Pull request closed without merging**